text/0098-generalized-templates.md

```diff
@@ -0,0 +1,277 @@
+
+By using generalized templates compared to an attribute of an entity to express this policy, we do not have to store information that is only relevant to our authorization decisions within an external data store. This allows us to have a cleaner seperation of concerns. 
+
+#### Schema 
+```
+entity Document = {
+    owner: Person 
+};
+entity Person;
+
+action Access appliesTo {
+    principal: FS::Person,
+    resource: [FS::Document],
+    context: { date: datetime }
+};
+```
+
+#### Template 
+```
+template(?startTime: datetime, ?endTime: datetime) =>
+permit(principal == ?principal, 
+action, 
+resource == ?resource) when 
+{ ?startTime <= context.now && 
+context.now <= ?endTime };
+```
+
+### Example 2
+In this example, suppose we have a file sharing application where we can create folders and documents. Every document created must be in a folder. We want to express the policy that if you are an owner of the folder then you can access any document within it. We can use generalized templates to define this as shown below. 
+
+Now everytime we create a folder, we link ```?resource``` with that folder to express this relationship. 
+
+An alternative way to do this is to store an additional owner attribute for every document, however this has a couple of drawbacks compared to using Generalized Templates. 
+
+(1) We lose the intent behind why we are authoring this policy. 
+
+(2) When transferring ownership of a folder to someone else you would have to manually update the owner attribute for all document's that belong to that folder 
+
+#### Schema
+```
+namespace FS {
+    entity Folder;
+    entity Document in Folder;
+    entity Person = {
+        owned_folders: Set<Folder>
+    };
+}
+
+action AccessDocument appliesTo {
+    principal: FS::Person,
+    resource: FS::Document,
+};
+```
+
+#### Template
+```
+permit(principal, action == Action::"AccessDocument", resource in ?resource) when { 
+    ?resource in principal.owned_folders
+};
+```
+
+### Example 3
+
+Suppose there are certain professors with dual appointments under two different departments who want to give students of their group access to both the resources in the first department and the resources in their second department. 
+
+Note: Although we can create two seperate templates and instantiate them individually, we would need to make sure they stay in sync. With generalized templates, since there is only one template we do not need to do so. Error states where only one part of the template is applied but not the other would not be able to occur. Generalized templates allows us to subsume the functionality described in [#7](https://github.com/cedar-policy/rfcs/pull/7) and with typed slots we can express the motivating example for template-groups.
```

Comment:
```suggestion
Note: Although we can create two separate templates and instantiate them individually, we would need to make sure they stay in sync. With generalized templates, since there is only one template we do not need to do so. Error states where only one part of the template is applied but not the other would not be able to occur. Generalized templates allows us to subsume the functionality described in [#7](https://github.com/cedar-policy/rfcs/pull/7) and with typed slots we can express the motivating example for template-groups.
```